### ISSUE TITLE
Request a new padding mode

### ISSUE THREAD

**Issue by ppwwyyxx** [tensorflow/tensorflow#18213]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: no
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: n/a
- **TensorFlow installed from (source or binary)**: n/a
- **TensorFlow version (use command below)**: n/a
- **Python version**: n/a
- **Bazel version (if compiling from source)**: n/a
- **GCC/Compiler version (if compiling from source)**: n/a
- **CUDA/cuDNN version**: n/a
- **GPU model and memory**: n/a
- **Exact command to reproduce**: n/a

According to https://www.tensorflow.org/api_guides/python/nn#Convolution, the current padding mode "SAME" will depend on the input size to determine how many pixels to pad, for example:

input=225, kernel=7, stride=2  ---> padding = [3, 3]
input=224, kernel=7, stride=2  ---> padding = [2, 3]

However, in most other CNN implementations, (and also, historically), padding does not depend on the input size. For kernel=7 and stride=2, padding usually will be [3, 3] (which is actually equivalent to [3, 2] when input=224).

Potential issues:
1. Inconsistent with models trained in other frameworks. It's not the first time I have to manually fix the padding when loading a model released by others, e.g. [here](https://github.com/ppwwyyxx/tensorpack/blob/1139854d7e286b56f87a92f96fe8f1b70789d794/examples/ResNet/load-resnet.py#L40-L42). This also causes pain for multi-backend framework such as Keras, because "SAME" does not mean the same thing for each backend. One example Keras issue [here](https://github.com/keras-team/keras/pull/9473). Also Keras has to explicit pad the image in its ResNet50 model: [here](https://github.com/keras-team/keras/blob/ef13db05731bfd53fa0a877637c99c1734be933b/keras/applications/resnet50.py#L213).

2. Due to how padding is computed (by `left=total_padding//2, right=total_padding-left`), the number of pixels padded on __left or top__ of the image may change with different input size, as shown by the example above. This is not a good default and in particular harmful for pixel-level tasks, such as detection&segmentation, where all the annotations have an offset starting from the top-left corner of the image.

In fact, many of Google's own code has to fix this manually by `tf.pad`, for example:
1. tensorflow/benchmarks has a new mode called "SAME_RESNET": https://github.com/tensorflow/benchmarks/blob/bab8a61aaca3d2b94072ae2b87f0aafe1797b165/scripts/tf_cnn_benchmarks/convnet_builder.py#L183-L199
2. The recent tpu training code has a function called "conv2d_fixed_padding": https://github.com/tensorflow/tpu/blob/b189540102d6b5b40b1730d7e5ad5c884bae323c/models/experimental/resnet_bfloat16/resnet_model.py#L102-L107
3. slim has a function called `conv2d_same`: https://github.com/tensorflow/models/blob/696b69a498b43f8e6a1ecb24bb82f7b9db87c570/research/slim/nets/resnet_utils.py#L77-L122
4. google-research/big_transfer: https://github.com/google-research/big_transfer/blob/49afe42338b62af9fbe18f0258197a33ee578a6b/bit_tf2/models.py#L36-L38
5. google-research/simclr: https://github.com/google-research/simclr/blob/6bf69ce127ae33e181e1a6c5777c84570cb5d147/tf2/resnet.py#L160-L183
6. TPU object detection codebase: https://github.com/tensorflow/tpu/blob/6f9c87c1215a67fd97da7272eff247e53f266e80/models/official/detection/modeling/architecture/nn_ops.py#L309-L327
Given all these I think it's reasonable to add a new mode to make things easier.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 21 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ppwwyyxx** [comment]
@reedwm sorry for @ing random people that are active on github. Could you raise this issue to the right person at TF's team for some comments?

**Comment by reedwm** [comment]
There is an internal bug assigned to me to allow for explicit padding. After explicit padding is implemented, it would be easy to add other other padding modes like a `SAME_RESNET` mode. Although I'm not sure the API team would want to add more padding modes.

The reason I haven't prioritized explicit padding is that I saw no performance gain in tf_cnn_benchmarks of using explicit padding to conv2d over using `tf.pad` followed by conv2d. It's relatively easy for users to call `tf.pad`. So the only benefit of adding more modes would be a slight usability improvement.

/CC @martinwicke thoughts on adding a new padding mode for consistency with other frameworks? This can be done even before I implement explicit padding.

**Comment by ppwwyyxx** [comment]
Thanks! Looks like I found the right person.
Explicit padding would be nice. Surprised to know that there is no gain over `tf.pad`. Speaking of performance, another potential issue with the current "SAME" mode is that it does padding twice (one explicitly with `functor::PadInput` and one with cudnn) in order to pad [2, 3]. But in the hypothetical new mode, only one cudnn call is needed to pad [3, 2] .

**Comment by reedwm** [comment]
`functor::PadInput` will still be needed in some cases. cudnn only allows equal padding on both sides. It is possible to sometimes allow the left-padding to be slightly larger than the right-padding with cudnn, so in some cases, we might not need `functor::PadInput` when the padding is [3, 2]. But, when I measured the performance of `tf_cnn_benchmarks` with explicit padding, I didn't implement that optimization, and always would call `functor::PadInput` when the left-padding was larger than the right-padding.

**Comment by martinwicke** [comment]
Once @reedwm's explicit padding is in, adding another constant to the API is not a problem. We should call it something other than "SAME_RESNET", maybe "VALID_STATIC"?

Before the fix, adding it may be too disruptive to be worthwhile, but I'll defer to Reed on that.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @shivaniag: It has been 57 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by reedwm** [comment]
Currently working on explicit padding, which is a prerequisite to adding a new padding mode.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @reedwm: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by davideboschetto** [comment]
@reedwm This is huge, it's good to know a fix is actively under development! It's a mess, one supposes that "same" means the same thing across all frameworks!

**Comment by tensorflowbutler** [comment]
Nagging Assignee @reedwm: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @reedwm: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @reedwm: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @reedwm: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @reedwm: It has been 76 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @reedwm: It has been 91 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by zjzh** [comment]
> Currently working on explicit padding, which is a prerequisite to adding a new padding mode.
@reedwm, Hello, why is it necessary to adding explicit padding in convolution. Is it because that usability improvement and other frameworks support?

**Comment by reedwm** [comment]
> why is it necessary to adding explicit padding in convolution

Explicit padding is now implemented in several ops, including Conv2D.  Supporting explicit padding means its easy to add support for a new padding mode, as the op will already support arbitrary padding.

It is now relatively easy to add a new padding mode, but I am not sure if it is worth the API change. Every new padding mode means more complexity when a user tries to understand the `padding` parameter, and its already possible to manually implementing whatever padding algorithm you wish with explicit padding (or `tf.pad`).

@martinwicke do you think a new padding mode should be added?

**Comment by tilakrayal** [comment]
@ppwwyyxx,
Have you got the chance to have a look at tf.nn.conv2d  where the explicit padding was implemented.
https://www.tensorflow.org/api_docs/python/tf/nn/conv2d

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Thank you!
